File: backend/src/entities/PointTransaction.ts
import {
  Entity, PrimaryGeneratedColumn, Column, CreateDateColumn,
  ManyToOne, JoinColumn,
} from 'typeorm';
import { PointReason } from '../types/index.js';
import { Participant } from './Participant.js';
import { Event } from './Event.js';

@Entity('point_transactions')
export class PointTransaction {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ type: 'int' })
  amount: number;

  @Column({ type: 'int' })
  balanceAfter: number;

  @Column({ type: 'enum', enum: PointReason })
  reason: PointReason;

  @Column({ length: 255, nullable: true })
  note: string | null;

  @Column({ length: 100, nullable: true })
  pointType: string | null;

  @Column('uuid', { nullable: true })
  givenBy: string | null;

  @Column({ type: 'text', nullable: true })
  signature: string | null;

  @ManyToOne(() => Participant, (p) => p.transactions, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'participantId' })
  participant: Participant;

  @Column('uuid')
  participantId: string;

  @ManyToOne(() => Event, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'eventId' })
  event: Event;

  @Column('uuid')
  eventId: string;

  @CreateDateColumn()
  createdAt: Date;
}

File: app/lib/models/point_transaction.dart
class PointTransaction {
  final String id;
  final int amount;
  final int balanceAfter;
  final String reason;
  final String? note;
  final String? pointType;
  final String participantId;
  final String createdAt;

  const PointTransaction({
    required this.id,
    required this.amount,
    required this.balanceAfter,
    required this.reason,
    this.note,
    this.pointType,
    required this.participantId,
    required this.createdAt,
  });

  factory PointTransaction.fromJson(Map<String, dynamic> json) =>
      PointTransaction(
        id: json['id'] as String,
        amount: json['amount'] as int,
        balanceAfter: json['balanceAfter'] as int,
        reason: json['reason'] as String,
        note: json['note'] as String?,
        pointType: json['pointType'] as String?,
        participantId: json['participantId'] as String? ?? '',
        createdAt: json['createdAt'] as String,
      );
}
